includes/sanitizers/class-amp-style-sanitizer.php

```diff
@@ -2138,7 +2138,25 @@ private function process_css_list( CSSList $css_list, $options ) {
 					$this->process_css_declaration_block( $css_item, $css_list, $options )
 				);
 			} elseif ( $css_item instanceof AtRuleBlockList ) {
-				if ( ! in_array( $css_item->atRuleName(), $options['allowed_at_rules'], true ) ) {
+				if (
+					'-moz-document' === $css_item->atRuleName()
+					&&
+					'url-prefix()' === $css_item->atRuleArgs()
+					&&
+					in_array( 'supports', $options['allowed_at_rules'], true )
+				) {
+					// Replace `@-moz-document url-prefix()` with `@supports (-moz-appearance:meterbar)` as an alternative
+					// way to provide Firefox-specific style rules. This is a workaround since @-moz-document is not
+					// yet allowed in AMP, and this use of @supports is another recognized Firefox-specific CSS hack,
+					// per <http://browserhacks.com/#hack-8e9b5504d9fda44ec75169381b3c3157>.
+					// For adding @-moz-document to AMP, see <https://github.com/ampproject/amphtml/issues/26406>.
+					$new_css_item = new AtRuleBlockList( 'supports', '(-moz-appearance:meterbar)' );
+					$new_css_item->setContents( $css_item->getContents() );
+					$css_list->replace( $css_item, $new_css_item );
+					$css_list->remove( $css_item ); // @todo This should be redundant with the replace() above.
```

Comment:
This should not be needed!

Reply:
This may be because of some `$css_item` contents being sanitized and removed prior to this specific item. If that's the case the index being used for splicing may be incorrect due to other contents being removed from the list. For example, with this modified test (prior to a4bf60a0bd5e668c323250d0eb633d7465528f4d):

```diff
--- a/tests/php/test-amp-style-sanitizer.php	(revision c772d292bda750774693c7a3399d2e870c92f2cc)
+++ b/tests/php/test-amp-style-sanitizer.php	(date 1603383215756)
@@ -202,7 +202,7 @@
 			],
 
 			'allowed_at_rules_retained' => [
-				'<style>@media screen and ( max-width: 640px ) { body { font-size: small; } } @font-face { font-family: "Open Sans"; src: url("/fonts/OpenSans-Regular-webfont.woff2") format("woff2"); } @-moz-document url-prefix() { body { color:red; } } @supports (display: grid) { div { display: grid; } } @-moz-keyframes appear { from { opacity: 0.0; } to { opacity: 1.0; } } @keyframes appear { from { opacity: 0.0; } to { opacity: 1.0; } }</style><div></div>',
+				'<style>@charset "UTF-8"; @media screen and ( max-width: 640px ) { body { font-size: small; } } @font-face { font-family: "Open Sans"; src: url("/fonts/OpenSans-Regular-webfont.woff2") format("woff2"); } @-moz-document url-prefix() { body { color:red; } } @supports (display: grid) { div { display: grid; } } @-moz-keyframes appear { from { opacity: 0.0; } to { opacity: 1.0; } } @keyframes appear { from { opacity: 0.0; } to { opacity: 1.0; } }</style><div></div>',
 				'<div></div>',
 				[
 					'@media screen and ( max-width: 640px ){body{font-size:small}}@font-face{font-family:"Open Sans";src:url("/fonts/OpenSans-Regular-webfont.woff2") format("woff2")}@supports (-moz-appearance:meterbar){body{color:red}}@supports (display: grid){div{display:grid}}@-moz-keyframes appear{from{opacity:0}to{opacity:1}}@keyframes appear{from{opacity:0}to{opacity:1}}',
```

The `@charset` rule is sanitized and removed from the CSS list, and the `@-moz-document` rule is not replaced:

```
There was 1 failure:

1) AMP_Style_Sanitizer_Test::test_body_style_attribute_sanitizer with data set "allowed_at_rules_retained" ('<style>@charset "UTF-8"; @med...</div>', '<div></div>', array('@media screen and ( max-width...ty:1}}'))
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
 Array (
-    0 => '@media screen and ( max-width: 640px ){body{font-size:small}}@font-face{font-family:"Open Sans";src:url("/fonts/OpenSans-Regular-webfont.woff2") format("woff2")}@supports (-moz-appearance:meterbar){body{color:red}}@supports (display: grid){div{display:grid}}@-moz-keyframes appear{from{opacity:0}to{opacity:1}}@keyframes appear{from{opacity:0}to{opacity:1}}'
+    0 => '@media screen and ( max-width: 640px ){body{font-size:small}}@font-face{font-family:"Open Sans";src:url("/fonts/OpenSans-Regular-webfont.woff2") format("woff2")}@-moz-document url-prefix(){body{color:red}}@supports (-moz-appearance:meterbar){body{color:red}}@-moz-keyframes appear{from{opacity:0}to{opacity:1}}@keyframes appear{from{opacity:0}to{opacity:1}}'
 )
```

Reply:
I've committed that `@charset` change to the test data.

Nevertheless, the `replace` and `remove` methods are [as follows](https://github.com/sabberworm/PHP-CSS-Parser/blob/a525c75db322fd295371ba75df2aece58b70b895/lib/Sabberworm/CSS/CSSList/CSSList.php#L224-L253):

```php
	/**
	 * Removes an item from the CSS list.
	 * @param RuleSet|Import|Charset|CSSList $oItemToRemove May be a RuleSet (most likely a DeclarationBlock), a Import, a Charset or another CSSList (most likely a MediaQuery)
	 * @return bool Whether the item was removed.
	 */
	public function remove($oItemToRemove) {
		$iKey = array_search($oItemToRemove, $this->aContents, true);
		if ($iKey !== false) {
			unset($this->aContents[$iKey]);
			return true;
		}
		return false;
	}

	/**
	 * Replaces an item from the CSS list.
	 * @param RuleSet|Import|Charset|CSSList $oItemToRemove May be a RuleSet (most likely a DeclarationBlock), a Import, a Charset or another CSSList (most likely a MediaQuery)
	 */
	public function replace($oOldItem, $mNewItem) {
		$iKey = array_search($oOldItem, $this->aContents, true);
		if ($iKey !== false) {
			if (is_array($mNewItem)) {
				array_splice($this->aContents, $iKey, 1, $mNewItem);
			} else {
				array_splice($this->aContents, $iKey, 1, array($mNewItem));
			}
			return true;
		}
		return false;
	}
```

So in both cases, `array_search()` is being used to find the `$css_item`. But the `array_splice()` with the index doesn't seem to be having the desired effect.

Check this out: it looks like `array_splice()`'s use of `$offset` is _not_ as an index: https://3v4l.org/L8qQo

Reply:
Hah, I already encountered this problem before:

https://github.com/ampproject/amp-wp/blob/4cd3e06c1433e65382bb36f2caa3c5344e943356/includes/sanitizers/class-amp-style-sanitizer.php#L1754-L1755

Reply:
There we go. Issue is fixed in d7c565c8be4f71abb53328c44c96266103c6b7f4 along eliminating dependence on buggy `CSSList::replace()` and `CSSList::splice()`.

Reply:
Nice! Works perfectly now :100:.